labcontrol/db/process.py

```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
+            for control (blank/positive control/etc) samples.
+        plate_col_name: str
+            The name of the column in input_df that contains the name of the
+            plate on which a given sample lies.
+        projname_col_name: str
+            The name of the column in input_df that contains the name of the
+            project associated with the given sample.
+
+        Returns
+        -------
+        input_df: pandas.DataFrame
+            The input dataframe, modified so that the controls have the same
+            (single) project name as the experimental samples on their sample
+            plate.
+
+        Raises
+        ------
+        ValueError
+            If any plate contains experimental samples from more (or fewer)
+            than one project.
+        """
+
+        problem_plate_messages = []
+
+        # create a mask to define all the NON-control rows for this plate
+        non_controls_mask = input_df[projname_col_name].notnull()
+
+        # get all the unique plates in the dataframe
+        unique_plates = input_df[plate_col_name].unique()
+        for curr_unique_plate in unique_plates:
+            # create a mask to define all the rows for this plate
+            plate_mask = input_df[plate_col_name] == curr_unique_plate
+
+            # create a mask to define all the rows for this plate where the
+            # project name is NOT the control value (None)
+            plate_non_controls_mask = plate_mask & non_controls_mask
+
+            # get unique project names for the part of df defined in the mask
+            curr_unique_projnames = (input_df[plate_non_controls_mask]
+                                  [projname_col_name].unique())
+            if len(curr_unique_projnames) != 1:
+                # Note that we don't error out the first time we find a
+                # plate that doesn't meet expectations; instead we continue to
+                # run through all the plates and identify ALL those that don't
+                # meet expectations.  This way the user can correct all of them
+                # at once.
+                curr_err_msg = "Expected one unique value for plate '{0}' " \
+                               "but received {1}: {2}".format(
+                    curr_unique_plate, len(curr_unique_projnames),
+                    ", ".join([str(x) for x in curr_unique_projnames]))
+                problem_plate_messages.append(curr_err_msg)
+            else:
+                # create a mask to define all the rows for this plate where the
+                # projname IS the control value (None); ~ "nots" a whole series
+                plate_controls_mask = plate_mask & (~non_controls_mask)
+
+                # ok to just take first non-control projname because we
+                # verified above there is only one value there anyway
+                input_df.loc[plate_controls_mask, projname_col_name] = \
+                    curr_unique_projnames[0]
```

Comment:
....however, if `len(curr_unique_projnames) == 0`, this would throw an `IndexError`. 

If I'm thinking about this correctly, I believe this edge case could result if a plate entirely of controls is run.

Reply:
Nope, it won't throw an IndexError in this case--that situation is already taken care of above with the check `len(curr_unique_projnames) != 1` ... if the length is zero, you end up in the error handling and not ever in the `else`.  I have a test to verify this; will check it in in a moment :)

Reply:
Oh, wow, sorry. Yes, I misread the flow of logic there.